src/order/bounded.lean

```diff
@@ -0,0 +1,365 @@
+/-
+Copyright (c) 2022 Violeta Hernández Palacios. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Violeta Hernández Palacios
+-/
+import .min_max
+import .rel_classes
+import data.set.intervals.basic
+
+/-!
+# Bounded and unbounded sets
+
+We prove miscellaneous lemmas about bounded and unbounded sets. Many of these are just variations on
+the same ideas, or similar results with a few minor differences. The file is divided into these
+different general ideas.
+-/
+
+variables {α : Type*} {r : α → α → Prop} {s t : set α}
+
+/-! ### Subsets of bounded and unbounded sets -/
+
+theorem bounded_of_subset_bounded (hst : s ⊆ t) (hs : bounded r t) : bounded r s :=
+hs.imp $ λ a ha b hb, ha b (hst hb)
+
+theorem unbounded_of_unbounded_subset (hst : s ⊆ t) (hs : unbounded r s) : unbounded r t :=
+λ a, let ⟨b, hb, hb'⟩ := hs a in ⟨b, hst hb, hb'⟩
+
+/-! ### Alternate characterizations of unboundedness on orders -/
+
+lemma unbounded_le_of_forall_ex_lt [preorder α] (h : ∀ a, ∃ b ∈ s, a < b) : unbounded (≤) s :=
+λ a, let ⟨b, hb, hb'⟩ := h a in ⟨b, hb, λ hba, not_lt_of_ge hba hb'⟩
+
+lemma unbounded_le_iff [linear_order α] : unbounded (≤) s ↔ ∀ a, ∃ b ∈ s, a < b :=
+⟨λ h a, let ⟨b, hb, hba⟩ := h a in ⟨b, hb, lt_of_not_ge hba⟩, unbounded_le_of_forall_ex_lt⟩
+
+lemma unbounded_lt_of_forall_ex_le [preorder α] (h : ∀ a, ∃ b ∈ s, a ≤ b) : unbounded (<) s :=
+λ a, let ⟨b, hb, hb'⟩ := h a in ⟨b, hb, λ hba, not_le_of_gt hba hb'⟩
+
+lemma unbounded_lt_iff [linear_order α] : unbounded (<) s ↔ ∀ a, ∃ b ∈ s, a ≤ b :=
+⟨λ h a, let ⟨b, hb, hba⟩ := h a in ⟨b, hb, le_of_not_gt hba⟩, unbounded_lt_of_forall_ex_le⟩
+
+lemma unbounded_ge_of_forall_ex_gt [preorder α] (h : ∀ a, ∃ b ∈ s, b < a) : unbounded (≥) s :=
+λ a, let ⟨b, hb, hb'⟩ := h a in ⟨b, hb, λ hba, not_lt_of_ge hba hb'⟩
```

Comment:
```suggestion
@unbounded_le_of_forall_ex_lt (order_dual α) _ _ h
```
All lemmas in this file about `ge` or `gt` can be proven from the version about `le` or `lt` by explicitly telling Lean that you apply it to the type `order_dual α`

Reply:
That's amazing! I figured this much, but I wasn't able to get it to work.